packages/core/lib/console.js

```diff
@@ -89,12 +78,71 @@ class Console extends EventEmitter {
       return new Promise(() => {});
     } catch (error) {
       this.options.logger.log(
-        "Unexpected error: Cannot provision contracts while instantiating the console."
+        "Unexpected error setting up the environment or provisioning " +
+        "contracts while instantiating the console."
       );
       this.options.logger.log(error.stack || error.message || error);
     }
   }
 
+  hydrateUserDefinedVariables() {
+    if (!this.options.console || !this.options.console.require) return;
+
+    const requireFromPath = target => {
+      return path.isAbsolute(target) ?
+        require(target) :
+        require(path.join(this.options.working_directory, target));
+    };
+    const addToContext = (userData, namespace) => {
+      for (const key in userData) {
+        if (namespace) {
+          if (typeof this.repl.context[namespace] === "undefined") {
+            this.repl.context[namespace] = {};
+          }
+          this.repl.context[namespace][key] = userData[key];
+        } else {
+          this.repl.context[key] = userData[key];
+        }
+      }
+    };
+    const errorMessage = "You must specify the console.require property as " +
+      "either a string or an array. If you specify an array, its members " +
+      "must be paths or objects containing at least a `path` property.";
+
+    if (typeof this.options.console.require === "string") {
+      addToContext(requireFromPath(this.options.console.require));
+    } else if (Array.isArray(this.options.console.require)) {
+      this.options.console.require.forEach(item => {
+        if (typeof item === "string") {
+          addToContext(requireFromPath(item));
+        } else if (typeof item === "object" && item.path) {
+          addToContext(requireFromPath(item.path), item.as);
+        } else {
+          throw new Error(errorMessage);
+        }
+      });
+    } else {
+      throw new Error(errorMessage);
+    }
+  }
+
+  async setUpEnvironment() {
+    let accounts;
+    try {
+      accounts = await this.interfaceAdapter.getAccounts();
+    } catch {
+      // don't prevent Truffle from working if user doesn't provide some way
+      // to sign transactions (e.g. no reason to disallow debugging)
+      accounts = [];
+    }
+
+    this.repl.context.web3 = this.web3;
+    this.repl.context.interfaceAdapter = this.interfaceAdapter;
+    this.repl.context.accounts = accounts;
+
+    this.hydrateUserDefinedVariables();
```

Comment:
The current ordering would allow our globals to be over written. 

```suggestion
    this.hydrateUserDefinedVariables();
    
    this.repl.context.web3 = this.web3;
    this.repl.context.interfaceAdapter = this.interfaceAdapter;
    this.repl.context.accounts = accounts;

```

Reply:
Agreed that we probably want to make sure our globals are safe. We're giving a namespacing mechanism, anyway!

Reply:
I would say just allow it. Maybe they want their own version of web3 or something. These globals are, after all, for the user's convenience.

Reply:
We could document that Truffle will set the context of `web3`, `interfaceAdapter` and `accounts` after user scripts are loaded and recommend a pattern for overriding. For example, creating an `overrideTruffleGlobals(global)` snippet they invoke at the console prompt.

Reply:
I don't understand why we should care about this? Truffle itself doesn't use them. Let them make their own custom web3 or accounts if they want. Why is this a problem?

Reply:
This ordering gives us flexibility to insert more globals in the future without worrying about users snippets and breaking behavior.